### PR TITLE
OCPBUGS-43994: UPSTREAM: <carry>: kube-apiserver: Only log each HTTP request once

### PR DESCRIPTION
**Pull request by tchap**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

kube-apiserver now logs each HTTP request twice since there is an extra logging call when finishing the request. This duplicity is now removed.

#### Which issue(s) this PR fixes:

N/A

#### Special notes for your reviewer:

N/A

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

N/A

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go [hunk 1]
```diff
@@ -131,16 +131,15 @@ func withLogging(handler http.Handler, stackTracePred StacktracePred, shouldLogR
 		req = req.WithContext(context.WithValue(ctx, respLoggerContextKey, rl))
 
 		var logFunc func()
-		logFunc = rl.Log
+		if klog.V(3).Enabled() || (rl.isTerminating && klog.V(1).Enabled()) {
+			logFunc = rl.Log
+		}
 		defer func() {
 			if logFunc != nil {
 				logFunc()
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:140]: so, previously, the request was first logged here

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go [hunk 2]
```diff
@@ -131,16 +131,15 @@ func withLogging(handler http.Handler, stackTracePred StacktracePred, shouldLogR
...
 			}
 		}()
 
-		if klog.V(3).Enabled() || (rl.isTerminating && klog.V(1).Enabled()) {
-			defer rl.Log()
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:142]: and then the second time here, right ?

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:142]: do you have an example of a CI jobs that logs requests twice ?

Comment (tchap) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:142]: I don't have a CI job, I used the client case to identify the issue: https://access.redhat.com/support/cases/#/case/03954830

It seems our CI jobs use `--v=2`, which is not enough to see request logs.

Comment (tchap) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:142]: From the case:

> I tried with `--v=2` but this results in no logging of the requests. We need to log the requests for support and debug of issues, so will need you to "Pursue fixing the apiserver code so that we don't double log when --v=3 and the apiserver is not terminating."

> We need an option that logs each request once, whereas the only options now are 0 (with --v=2) or twice (with --v=3)

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go [hunk 3]
```diff
@@ -131,16 +131,15 @@ func withLogging(handler http.Handler, stackTracePred StacktracePred, shouldLogR
...
+		if klog.V(3).Enabled() || (rl.isTerminating && klog.V(1).Enabled()) {
+			logFunc = rl.Log
+		}
 		defer func() {
 			if logFunc != nil {
 				logFunc()
 			}
 		}()
 
-		if klog.V(3).Enabled() || (rl.isTerminating && klog.V(1).Enabled()) {
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:141]: I think this code relies on the ShouldLogRequestPred passed to the function. If the predicate doesn’t match, then nothing gets logged. Right?

By default, the predicate is set to log at v=3, so this code doesn’t really make sense to me. How about simply removing it?

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go (hunk 2 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:142]: So, how are we going to test the fix ?

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go (hunk 3 above)
Comment (tchap) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:141]: Oh yeah, you are absolutely right. I guess I focused too much on simply removing the extra log message 👍🏻

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go (hunk 2 above)
Comment (tchap) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:142]: I will add a test to verify this.

**Comment by p0lyn0mial** [comment]
@tchap please also find the pr that introduced this change and let's add a description to the commit pointing to the original pr with which the new changes should be squashed during the next rebase.

**Comment by tchap** [comment]
@p0lyn0mial Is this any kind of standardized process documented anywhere, or should I just add a generic comment?

**Comment by p0lyn0mial** [comment]
> @p0lyn0mial Is this any kind of standardized process documented anywhere, or should I just add a generic comment?

@bertinatto ^^

**Comment by bertinatto** [comment]
> > @p0lyn0mial Is this any kind of standardized process documented anywhere, or should I just add a generic comment?
> 
> @bertinatto ^^

We don't have a standarized process, but historically we either:

1) use the same commit title as the original commit that introduced the change;
2) or we add a `(squash)` to the commit title plus the original commit in the description (personally, I prefer this one).

That way, the person doing the rebase will easily notice that the carry patch can/should be squashed.

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go [hunk 4]
```diff
@@ -102,19 +101,24 @@ const withLoggingLevel = 3
 
 // WithLogging wraps the handler with logging.
 func WithLogging(handler http.Handler, pred StacktracePred, isTerminatingFn func() bool) http.Handler {
-	return withLogging(handler, pred, func() bool {
-		return klog.V(withLoggingLevel).Enabled()
-	}, isTerminatingFn)
+	return withLogging(handler, pred, func(ctx context.Context) bool {
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:105]: wait, why so many changes in this file ? should we simply remove the code we discussed previously ?

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go [hunk 5]
```diff
@@ -102,19 +101,24 @@ const withLoggingLevel = 3
...
+		isTerminating := false
+		if isTerminatingFn != nil {
+			isTerminating = isTerminatingFn()
+		}
+		logger := klog.FromContext(ctx)
+		return logger.V(withLoggingLevel).Enabled() || (isTerminating && logger.V(1).Enabled())
+	})
```
Comment (tchap) [inline staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:112]: @p0lyn0mial I ended up refactoring the logic and private interfaces a bit, but the overall approach is not exactly bullet-proof.

There is now a single predicate function that merges previous `shouldLogRequest` and `isTerminatingFn`. This means `withLogging` is now much more clear and simple.

I added tests and use contextual logging internally that is handy for testing.

The issue really is that when you check `respLogger.Log`, it contains anyway `klog.V(withLoggingLevel)`, but `isTerminating` flag is not passed there, so I am afraid the current implementation contains a bug anyway and `isTerminating` is pretty much ignored. So what I did was to check all the conditions initially here and remove `.V` call from `.Log`. This makes all the tests pass just fine.

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go [hunk 6]
```diff
@@ -268,7 +261,7 @@ func SetStacktracePredicate(ctx context.Context, pred StacktracePred) {
 }
 
 // Log is intended to be called once at the end of your request handler, via defer
-func (rl *respLogger) Log() {
+func (rl *respLogger) Log(ctx context.Context) {
```
Comment (tchap) [inline staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:271]: There is a change of interface, but it doesn't seem to be really used publicly as this is an unexported object.

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go (hunk 5 above)
Comment (tchap) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:112]: Another approach would be to keep the `isTerminating` field in the logger and repeat the terminating check in `Log` as well...

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:112]: I need to check but if `rl.isTerminating` is/was unused then maybe we can drop the entire patch. Assuming our issue was:

```
if klog.V(3).Enabled() || (rl.isTerminating && klog.V(1).Enabled()) {...}
```

Comment (tchap) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:112]: This `isTerminatingFunc` is actually set up and passed to this machinery in https://github.com/openshift/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/config.go#L1155

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:112]: yeah, I think that the patch was introduced in https://github.com/openshift/kubernetes/commit/b052cdedb7bbd1683f213deb5c74cee9d278168b

But given that shouldLogRequest is always true when the loglevel is 3 (hardcoded value) and the fact that rl.isTerminating is unused beyond httplog what is the point of the original patch/pr ?

If we were to drop it, I think there would be no change in the behavior. am I right?

Comment (tchap) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:112]: I think that the original patch is simply wrong and it's never been really tested.

The original idea was, I assume, to enable logging when terminating. But the only effect actually is that all requests are logged twice when on `V(3)`.

So the `V(1)` check is useless and `isTerminating` as well unless we are at `V(3)` anyway. If we wanted to make this work, we would actually need to move the check to `.Log` where the actually logging happens, and there it would really work to use

```golang
if klog.V(3).Enabled() || (isTerminating && klog.V(1).Enabled() {
    klog.Infof(...) // Notice no V() here.
}
```

I will prepare one last version if we want to enable logging when terminating that is actually correct and simple. Otherwise we can just drop the original patch, which would mean no logging when terminating. Which is not happening right now anyway, but the question is whether we want to drop the intent.

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:112]: > Otherwise we can just drop the original patch, which would mean no logging when terminating. Which is not happening right now anyway, but the question is whether we want to drop the intent.

I think the patch is very old. I think that the purpose was to find requests during server termination - not a feature users were using. 

I think that we could simply drop the patch and test manually that when loglevel is 3 there are no duplicates during server shutdown.

Comment (tchap) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:112]: I did push one last commit, but I guess the plan is to close this and also remove the original carry this is fixing then?

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:112]: I think so, given that the behaviour will not change after removing the PR, right?. Unless there are some other parts that depend on the original PR.

Comment (tchap) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:112]: I don't think anything depends on it, I've reverted the commit just to test it.

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:112]: could you also test manually that httplog logs requests when appropriate loglevel is set and there are no duplicates ?

you can also open a new PR so that we get a signal from the CI.

**Comment by tchap** [comment]
We can close this as we decided to drop the carry: https://github.com/openshift/kubernetes/pull/2477
